README.md
# foppl

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar foppl-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-10-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-10-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/foppl/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/foppl/compare/0.1.0...0.1.1


src/foppl/primitives.clj
(ns foppl.primitives
  (:require [anglican.runtime :refer :all])
  (:require [anglican.core :refer :all])
  (:require [clojure.core.matrix :as m])
  (:require [anglican.runtime :refer [tanh]]))

(def ^:dynamic *primitive-procedures*
  "primitive procedures, do not exist in CPS form"
  (let [;; higher-order procedures cannot be primitive
        exclude '#{loop
                   map reduce
                   filter keep keep-indexed remove
                   repeatedly
                   every? not-any? some
                   every-pred some-fn
                   comp juxt partial}
        ;; runtime namespaces
        runtime-namespaces '[clojure.core anglican.runtime foppl.primitives]]
    (set (keep (fn [[k v]]
                 (when (and (not (exclude k))
                            (fn? (var-get v)))
                   k))
               (mapcat ns-publics runtime-namespaces)))))

  (defn append [& args] (apply conj args))

  (defn mat-mul [& args] (apply m/mmul args))
  (defn mat-add [& args] (apply m/add args))
  (defn mat-transpose [& args] (apply m/transpose args))
  (defn mat-tanh [M] (m/emap tanh M))
  (defn mat-relu [M] (m/emap (fn [x] (if (> x 0) x 0)) M))
  (defn mat-repmat [M r c]
    (let [R (reduce (partial m/join-along 0) (repeat r M))]
      (reduce (partial m/join-along 1) (repeat c R))))

(def env (hash-map "normal" #'normal
                   "sqrt" #'sqrt
                   "uniform-continuous" #'uniform-continuous
                   "beta" #'beta
                   "bernoulli" #'bernoulli
                   "exponential" #'exponential
                   "discrete" #'discrete
                   "append" #'foppl.primitives/append
                   "+" #'+
                   "-" #'-
                   "*" #'*
                   "/" #'/
                   "exp" #'exp
                   "vector" #'vector
                   "last" #'last
                   "nth" #'nth
                   "hash-map" #'hash-map
                   "put" #'assoc
                   "get"  #'get
                   "last" #'last
                   "first" #'first
                   "second" #'second
                   "rest" #'rest
                   "cons" #'cons
                   "conj" #'conj
                   "mat-add" #'mat-add
                   "mat-mul" #'mat-mul
                   "mat-repmat" #'mat-repmat
                   "mat-tanh" #'mat-tanh
                   "mat-transpose" #'mat-transpose
                   "sample*" #'sample*
                   "observe*" (fn [& args] nil)
                   ))


src/foppl/graph_based_sampling.clj
(require '[clojure.data.json :as json])
(require '[clojure.core.match :refer [match]])
(require '[anglican.runtime :refer :all])
(load "primitives")
(require '[clojure.java.shell :as shell :refer [sh]])
(use '[clojure.java.shell :only [sh]])

(def all-records (json/read-str (slurp (str (cwd) "../CS532-HW2/graph_test_2.json"))
                :key-fn str))

(defn deterministic-eval [exp]
  (match [exp]
         [([h & t] :seq)] (apply (foppl.primitives/env h) (map deterministic-eval t))
         [d] (double d)
         :else (throw (Exception. (str "Expression type unknown." exp)))
         ))

(defn probabilistic-eval [exp var-map]
  (match [exp]
         [([h & t] :seq)]
         (apply (foppl.primitives/env h) (map (fn [e] (probabilistic-eval e var-map)) t))
         [d] (if (or (= (type d) java.lang.Long) (= (type d) java.lang.Double)) d
                 (if (= (type d) java.lang.String) (var-map d)
                    (throw (Exception. (str "Expression type unknown." d)))
         ))))

(defn call-prob-eval [var-map var graph]
  (let [link-fun (((second graph) "P") var)]
    (assoc var-map var (probabilistic-eval link-fun var-map))
  ))

(defn get-parents-keys [var parents akeys A]
  (if (empty? akeys) parents
    (let [key (first akeys) tail (rest akeys)]
      (if (some #(= var %) (A key)) (get-parents-keys var (foppl.primitives/append parents key) tail A) (get-parents-keys var parents tail A))
  )))

(defn get-parents-from-dict [var A]
  (get-parents-keys var '() (keys A) A))

(defn get-parents [var graph]
  (get-parents-from-dict var ((second graph) "A")))

(defn top-sort [queue in-q vars G]
  (if (empty? vars) [queue in-q]
    (let [var (first vars) tail (rest vars)]
      (if (get in-q var false) (top-sort queue in-q tail G)
        (let [parents (get-parents var G) new-queues (top-sort queue in-q parents G)]
          (top-sort (foppl.primitives/append (first new-queues) var)
                    (assoc (last new-queues) var true) tail G)))
)))

(defn top-sort-graph [G]
  (top-sort [] {} (keys ((second G) "P")) G))

(defn sample-from-joint [graph]
  (let [queues (top-sort-graph graph)
        vars (first queues)
        samples (reduce
                 (fn [new-map var] (call-prob-eval new-map var graph)) {} vars)
        return (probabilistic-eval (last graph) samples)]
    return
  ))

(defn diff [ret truth]
  (if (= (type truth) clojure.lang.PersistentVector)
    (reduce + (map - ret truth))
    (- ret truth)))

(defn cwd [] (let [pwd ((sh "pwd") :out)] (str (.substring pwd 0 (- (count pwd) 1)) "/")))
(defn graph-tests [type i]
  (json/read-str ((shell/with-sh-dir
    (str (cwd) "../daphne")
    (sh "lein" "run" "-f" "json" "desugar" "-i" (str "../CS532-HW2/programs/tests/" type "/test_" i ".daphne"))) :out)))

(defn run-deterministic-tests []
  (for [i (map (fn [x] (+ 1 x)) (range 13))]
    (do (println (str "running test " i))
    (let [ast (graph-tests "deterministic" i)
          truth (load-truth i)
          [ret sig] (evaluate-program ast)]
            (if (> (abs (diff truth ret)) 0.001)
              (let [msg (str "return value " ret " is not equal to truth " truth " for exp " ast)]
                    (do (print msg) msg)
              (do (print "passed test " i) (str "passed test " i))))))))

(run-deterministic-tests)


src/foppl/evaluation_based_sampling.clj
(require '[clojure.data.json :as json])
(require '[clojure.core.match :refer [match]])
(load "primitives")
(require '[clojure.java.shell :as shell :refer [sh]])
(use '[clojure.java.shell :only [sh]])
(use '(incanter core stats charts io))

(def rho {})
(defn evaluate [e s l]
  (match [e]
         [["sample" d]] (let [[dist sig] (evaluate d s l)
                               smpl-fun (foppl.primitives/env "sample*")]
                                   [(smpl-fun dist) sig])
         [["observe" d v]] nil
         [["let" [v1 e1] e0]] (let [[c1 sig] (evaluate e1 s l)]
                                (evaluate e0 sig (assoc l v1 c1)))
         [["if" e1 e2 e3]] (let [[e1' sig] (evaluate e1 s l)]
                                (if e1' (evaluate e2 sig l) (evaluate e3 sig l)))
         ; TODO cs isn't strictly correct since I am getting rid of the sigmas
         ; unrelated to the above, we check if f is in primitives: if so, call it
         ; using a primitive, if not, assume it was a user-defined procedure
         [([e0 & t] :seq)] (let [cs (map (fn [exp] (first (evaluate exp s l))) t)
                                 f (get foppl.primitives/env e0 false)]
                             (if f [(apply f cs) s] (let [[vs e0'] (rho e0)]
                                 (evaluate e0' s (merge l (zipmap vs cs))))))
         [nil] nil
         ; TODO if we add strings as primitives, then this will have to change
         ; checking if cv is not false is weird, but (get l c false) returns nil
         ; for variables bound to the output of observe statements, so I need it
         [c] (let [cv (get l c false)] (if (not (= cv false)) [cv s] [c s]))
         ))

(defn evaluate-program [ast]
  (match [ast]
         [[["defn" nm vs body] & t]] (do (def rho (merge rho {nm [vs body]})) (evaluate-program t))
         [[h]] (evaluate h [] {})))

(defn get-stream
  ([ast] (get-stream ast 1))
  ([ast n] (lazy-seq (cons (evaluate-program ast) (get-stream ast (inc n))))))

(defn diff [ret truth]
  (if (= (type truth) clojure.lang.PersistentVector)
    (reduce + (map - ret truth))
    (- ret truth)))

(defn cwd [] (let [pwd ((sh "pwd") :out)] (str (.substring pwd 0 (- (count pwd) 1)) "/")))
(defn desugar-tests [type i]
  (json/read-str ((shell/with-sh-dir
    (str (cwd) "../daphne")
    (sh "lein" "run" "-f" "json" "desugar" "-i" (str "../CS532-HW2/programs/tests/" type "/test_" i ".daphne"))) :out)))

(defn load-truth [type i]
  (load-file (str (cwd) "../CS532-HW2/programs/tests/" type "/test_" i ".truth")))

(defn run-deterministic-tests []
  (for [i (range 1 13)]
    (do (print (str "running test " i " "))
    (let [ast (desugar-tests "deterministic" i)
          truth (load-truth "deterministic" i)
          [ret sig] (evaluate-program ast)]
            (if (> (abs (diff truth ret)) 0.001)
              (let [msg (str "return value " ret " is not equal to truth " truth " for exp " ast)]
                    (do (println msg) msg))
                    (do (println "passed test " i) (str "passed test " i)))))))

;; (defn run-prob-test [samples]
;;   (let [
;;         sorted (sort samples)
;;         idx (range 1 (+ (count samples) 1))
;;         fracs (map (fn [x] (/ x (count samples))) idx)
;;         D+ (apply max (map - fracs sorted))
;;         D- (apply max (map - sorted (map (fn [x] (- x (/ 1 (count samples)))) fracs)))
;;         D (max D+ D-)
;;         ]
;;     (> D 0.565)
;;     ))

;; (defn run-probabilistic-tests []
;;   (let [num-samples (int 1e4) max_p_value 1e-4]
;;     (for [i (range 1 7)]
;;       (do (print (str "running test " i " "))
;;       (let [ast (desugar-tests "probabilistic" i)
;;             samples (map (fn [x] (first x)) (take num-samples (get-stream ast)))
;;             passed (run-prob-test samples)]
;;               (if passed
;;                 (do (println "failed test " i) (str "failed test " i))
;;                 (do (println "passed test " i) (str "passed test " i))))))))

;; (run-probabilistic-tests)

(run-deterministic-tests)

(defn desugar-programs [i]
  (json/read-str ((shell/with-sh-dir
    (str (cwd) "../daphne")
    (sh "lein" "run" "-f" "json" "desugar" "-i" (str "../CS532-HW2/programs/" i ".daphne"))) :out)))

(defn sample-task [i num-samples]
  (let [ast (desugar-programs i)
        samples (map first (take num-samples (get-stream ast)))]
    samples))

(defn divide [x d]
  (match [x]
         [[h]] (vector (div h d))
         [([h & t] :seq)] (cons (divide h d) (map (fn [x] (divide x d)) t))
         [a] (/ a d)))

(defn calc-expectation [samples]
  (if (= (type (first samples)) clojure.lang.PersistentVector)
    (map (fn [x] (divide x (count samples))) (apply foppl.primitives/mat-add samples))
    (divide (reduce + samples) (count samples))))

;; calculate expectations for tasks 1-3
(def mu (sample-task 1 1000))
(def slope-bias (sample-task 2 1000))
(def hmm (sample-task 3 1000))
(def nn (sample-task 4 1000))

;; create conditional histograms
(save (histogram mu) "mu.png")

(def slope (map (fn [x] (first x)) slope-bias))
(def bias (map (fn [x] (second x)) slope-bias))
(save (histogram slope) "slope.png")
(save (histogram bias) "bias.png")

(def hmm1 (map (fn [x] (first x)) hmm))
(save (histogram hmm1) "hmm1.png")
(def hmm6 (map (fn [x] (get x 6)) hmm))
(save (histogram hmm6) "hmm6.png")
(def hmm11 (map (fn [x] (get x 11)) hmm))
(save (histogram hmm11) "hmm11.png")
(def hmm16 (map (fn [x] (get x 16)) hmm))
(save (histogram hmm16) "hmm16.png")

(def nnw0 (map (fn [x] (first (first (first x)))) nn))
(save (histogram nnw0) "nnw0.png")
(def nnb0 (map (fn [x] (first (first (get x 1)))) nn))
(save (histogram nnb0) "nnb0.png")
(def nnw1 (map (fn [x] (first (first (get x 2)))) nn))
(save (histogram nnw1) "nnw1.png")
(def nnb1 (map (fn [x] (first (first (get x 3)))) nn))
(save (histogram nnb1) "nnb1.png")

;; calculate expectations of first three tasks
(calc-expectation mu)
(calc-expectation slope-bias)
(calc-expectation hmm)

;; calculate expectation of neural network task
(def first-sum (apply foppl.primitives/mat-add (map (fn [x] (first x)) nn)))
(def second-sum (apply foppl.primitives/mat-add (map (fn [x] (first (rest x))) nn)))
(def third-sum (apply foppl.primitives/mat-add (map (fn [x] (first (rest (rest x)))) nn)))
(def fourth-sum  (apply foppl.primitives/mat-add (map (fn [x] (first (rest (rest (rest x))))) nn)))
(def full-mat (vector first-sum second-sum third-sum fourth-sum))
(def expected-full-mat (div full-mat (count nn)))
